packages/swing-store-simple/test/test-state.js

```diff
@@ -72,3 +72,100 @@ test('rejectLMDB', t => {
   t.is(isSwingStore(notSimpleDir), false);
 });
+
+test('streamStore read/write', t => {
+  const { streamStore, commit, close } = initSwingStore();
+
+  let s1pos;
+  const writer1 = streamStore.openWriteStream('st1');
+  s1pos = writer1('first', s1pos);
+  s1pos = writer1('second', s1pos);
+  const s1posAlt = { ...s1pos };
+  const writer2 = streamStore.openWriteStream('st2');
+  s1pos = writer1('third', s1pos);
+  let s2pos = { itemCount: 0 };
+  s2pos = writer2('oneth', s2pos);
+  s1pos = writer1('fourth', s1pos);
+  s2pos = writer2('twoth', s2pos);
+  const s2posAlt = { ...s2pos };
+  s2pos = writer2('threeth', s2pos);
+  s2pos = writer2('fourst', s2pos);
+  streamStore.closeStream('st1');
+  streamStore.closeStream('st2');
+  const reader1 = streamStore.openReadStream('st1', s1pos);
+  const reads1 = [];
+  for (const item of reader1) {
+    reads1.push(item);
+  }
+  t.deepEqual(reads1, ['first', 'second', 'third', 'fourth']);
+  const writer2alt = streamStore.openWriteStream('st2');
+  s2pos = writer2alt('re3', s2posAlt);
+  streamStore.closeStream('st2');
+  const reader2 = streamStore.openReadStream('st2', s2pos);
+  const reads2 = [];
+  for (const item of reader2) {
+    reads2.push(item);
+  }
+  t.deepEqual(reads2, ['oneth', 'twoth', 're3']);
+
+  const reader1alt = streamStore.openReadStream('st1', s1pos, s1posAlt);
+  const reads1alt = [];
+  for (const item of reader1alt) {
+    reads1alt.push(item);
+  }
+  t.deepEqual(reads1alt, ['third', 'fourth']);
+
+  const writerEmpty = streamStore.openWriteStream('empty');
+  const emptyPos = writerEmpty('filler');
+  streamStore.closeStream('empty');
+  const readerEmpty = streamStore.openReadStream('empty', emptyPos, emptyPos);
+  const readsEmpty = [];
+  for (const item of readerEmpty) {
+    readsEmpty.push(item);
+  }
+  t.deepEqual(readsEmpty, []);
+  const readerEmpty2 = streamStore.openReadStream(
+    'empty',
+    streamStore.STREAM_START,
+  );
+  const readsEmpty2 = [];
+  for (const item of readerEmpty2) {
+    readsEmpty2.push(item);
+  }
+  t.deepEqual(readsEmpty2, []);
+
+  commit();
+  close();
+});
+
+test('streamStore mode interlock', t => {
+  const { streamStore, commit, close } = initSwingStore();
+
+  const writer = streamStore.openWriteStream('st1');
+  const s1pos = writer('first');
+  t.throws(() => streamStore.openReadStream('st1', s1pos), {
+    message: `can't read stream "st1" because it's already in use`,
+  });
+  t.throws(() => streamStore.openWriteStream('st1', s1pos), {
+    message: `can't write stream "st1" because it's already in use`,
+  });
+  streamStore.closeStream('st1');
+  t.throws(() => writer('second'), {
+    message: `can't write to closed stream "st1"`,
+  });
+
+  const reader = streamStore.openReadStream('st1', s1pos);
+  t.throws(() => streamStore.openReadStream('st1', s1pos), {
+    message: `can't read stream "st1" because it's already in use`,
+  });
+  t.throws(() => streamStore.openWriteStream('st1'), {
```

Comment:
I think this is the main case to guard against: the kernel (or the VatWarehouse) starts replaying a vat and is holding a read stream open, then fails to interlock properly and some crank gets delivered to the same vat, causing an attempted write to the transcript.